pkg/logql/rangemapper_test.go

```diff
@@ -0,0 +1,225 @@
+			`sum_over_time({app="foo"} | unwrap bar [3m])`,
+			`sum (downstream<sum_over_time({app="foo"} | unwrap bar [1m] offset 2m0s), shard=<nil>>
+				++ downstream<sum_over_time({app="foo"} | unwrap bar [1m] offset 1m0s), shard=<nil>>
+				++ downstream<sum_over_time({app="foo"} | unwrap bar [1m]), shard=<nil>>)`,
+			false,
+		},
+		{
+			`min_over_time({app="foo"} | unwrap bar [3m])`,
+			`min(downstream<min_over_time({app="foo"} | unwrap bar [1m] offset 2m0s), shard=<nil>>
+				++ downstream<min_over_time({app="foo"} | unwrap bar [1m] offset 1m0s), shard=<nil>>
+				++ downstream<min_over_time({app="foo"} | unwrap bar[1m]), shard=<nil>>)`,
+			false,
+		},
+		{
+			`min_over_time({app="foo"} | unwrap bar [3m]) by (bar)`,
+			`min by(bar)(downstream<min_over_time({app="foo"} | unwrap bar [1m] offset 2m0s) by (bar), shard=<nil>>
+				++ downstream<min_over_time({app="foo"} | unwrap bar [1m] offset 1m0s) by (bar), shard=<nil>>
+				++ downstream<min_over_time({app="foo"} | unwrap bar [1m]) by (bar), shard=<nil>>)`,
+			false,
+		},
+		{
+			`max_over_time({app="foo"} | unwrap bar [3m])`,
+			`max(downstream<max_over_time({app="foo"} | unwrap bar [1m] offset 2m0s), shard=<nil>>
+				++ downstream<max_over_time({app="foo"} | unwrap bar [1m] offset 1m0s), shard=<nil>>
+				++ downstream<max_over_time({app="foo"} | unwrap bar[1m]), shard=<nil>>)`,
+			false,
+		},
+		{
+			`max_over_time({app="foo"} | unwrap bar [3m]) by (bar)`,
+			`max by(bar)(downstream<max_over_time({app="foo"} | unwrap bar [1m] offset 2m0s) by (bar), shard=<nil>>
+				++ downstream<max_over_time({app="foo"} | unwrap bar [1m] offset 1m0s) by (bar), shard=<nil>>
+				++ downstream<max_over_time({app="foo"} | unwrap bar [1m]) by (bar), shard=<nil>>)`,
+			false,
+		},
+		// TODO: Add more binary operations
+		{
+			`rate({app="foo"}[3m]) + rate({app="bar"}[2m])`,
+			`(
+				sum(
+					downstream<rate({app="foo"}[1m] offset 2m0s), shard=<nil>>
+					++ downstream<rate({app="foo"}[1m] offset 1m0s), shard=<nil>>
+					++ downstream<rate({app="foo"}[1m]), shard=<nil>>
+				)
+				+
+				sum(
+					downstream<rate({app="bar"}[1m] offset 1m0s), shard=<nil>>
+					++ downstream<rate({app="bar"}[1m]), shard=<nil>>
+				)
+			)`,
+			false,
+		},
+		{
+			`min(rate({app="foo"}[5m])) / count(rate({app="bar"}[10m]))`,
+			`(
+				min(
+					downstream<min(rate({app="foo"}[1m] offset 4m0s)), shard=<nil>>
+					++ downstream<min(rate({app="foo"}[1m] offset 3m0s)), shard=<nil>>
+					++ downstream<min(rate({app="foo"}[1m] offset 2m0s)), shard=<nil>>
+					++ downstream<min(rate({app="foo"}[1m] offset 1m0s)), shard=<nil>>
+					++ downstream<min(rate({app="foo"}[1m])), shard=<nil>>
+				)
+				/
+				sum(
+					downstream<count(rate({app="bar"}[1m] offset 9m0s)), shard=<nil>>
+					++ downstream<count(rate({app="bar"}[1m] offset 8m0s)), shard=<nil>>
+					++ downstream<count(rate({app="bar"}[1m] offset 7m0s)), shard=<nil>>
+					++ downstream<count(rate({app="bar"}[1m] offset 6m0s)), shard=<nil>>
```

Comment:
nit: we may need `offset` support in regular LogQL queries to support actually downstreaming these to queriers and thus as a prerequisite to this PR.